### PR TITLE
Sentry performance integration. Provide `OrderDetails` transaction

### LINKED ISSUE #7056: [Orders Performance] Integrate Sentry Performance Monitoring
This issue is an epic for Sentry Performance Monitoring integration.

Tracks scope:

- https://github.com/Automattic/Automattic-Tracks-Android/pull/133
- https://github.com/Automattic/Automattic-Tracks-Android/pull/137
- https://github.com/Automattic/Automattic-Tracks-Android/pull/136

Woo scope:
- https://github.com/woocommerce/woocommerce-android/issues/7057

### LINKED ISSUE #7062: [Orders Performance] Introduce loading waiting time metric to the Orders details view

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #7056 
Relates to: #7062 

### Description

This PR is a draft of architecture for modeling performance tracking. To see, how things look under the Automattic-Tracks-Android library, please see the PR: https://github.com/Automattic/Automattic-Tracks-Android/pull/136

This PR is not yet ready for "normal" review but feel free to share ideas/concerns about architecture/design of the solution.

#### Key points I wanted to achieve
1. ViewModel doesn't know how performance tracking works. It only informs performance tracking that *something* happened.
2. Fragment is unaware of performance tracking. It only sends lifecycle events
3. Performance tracking doesn't know anything about Woo. It defines conditions when to start and when to finish tracking.

#### Debug variant

Let's start with `wasabiDebug` variant, as it has a default value of `1.0` sample rate of Performance Monitoring. It means that every transaction will be published to Sentry.

##### TC 1 Send transaction
1. Install the app
2. Open the app
3. Wait 5 seconds (*)
4. Open the app again (*)
5. Open any Order Details
6. - [ ] Check if OrderDetails transaction has been uploaded (up to 1 minute) (**)

##### TC 2 Do not send transaction if user disabled tracking
1. Install the app
2. Open the app
3. Wait 5 seconds (*)
4. Open the app again (*)
5. Go to settings
7. Disable user tracking (_not_ crash reporting)
8. Kill and re-open the app
9. Open any Order Details
10. - [ ] Check if OrderDetails transaction has **not** been uploaded (up to 1 minute) (**)

##### TC3
1. Install the app in variant `vanillaRelease`
2. Open the app
3. Wait 5 seconds (*)
4. Open the app again (*)
5. Open and close any Order Details. **Do this 10-20 times.**
6. - [ ] Check if OrderDetails transaction has been uploaded but there are less than 10-20 new transactions (up to 1 minute) (**)
We're testing if the release remote config has been fetched and assigned. We don't have direct control on Sentry sampling, but if there will be less than 10 events, sampling is different than `1.0`. This means that the config was fetched correctly.

> (*) Why do I have to do this?
We use Firebase Remote Config to set sample rate of performance monitoring. The default value is 0, and this value is assigned to Sentry initializer at the app start. So when the app starts for the first time, it will initialize with 0, and then fetch the remote config. For the second run, it will use the value from Firebase.

> (**) How to check uploaded transactions?
Go to Sentry dashboard -> Performance -> OrderDetails -> Filter: "Recent Transactions" -> click on `EVENT ID` for details.
Remember to set the correct environment (debug or release) at the top of the page.

### Images/gif

image

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/build.gradle [hunk 1]
```diff
@@ -15,7 +15,8 @@ plugins {
 
 sentry {
     tracingInstrumentation {
-        enabled = false
+        enabled = true
+        features = [io.sentry.android.gradle.extensions.InstrumentationFeature.DATABASE]
```
Comment (wzieba) [inline WooCommerce/build.gradle:19]: Other automated instrumentations are: `okhttp` (which we manually enable) and `file_io` which we don't need (at least now).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt [hunk 2]
```diff
@@ -98,9 +99,12 @@ class OrderDetailViewModel @Inject constructor(
     private val cardReaderTracker: CardReaderTracker,
     private val trackerWrapper: AnalyticsTrackerWrapper,
     private val shippingLabelOnboardingRepository: ShippingLabelOnboardingRepository,
+    private val orderDetailsTransactionLauncher: OrderDetailsTransactionLauncher,
 ) : ScopedViewModel(savedState), OnProductFetchedListener {
     private val navArgs: OrderDetailFragmentArgs by savedState.navArgs()
 
+    val performanceObserver: LifecycleObserver = orderDetailsTransactionLauncher
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt:106]: I want to expose `OrderDetailsTransactionLauncher` as `LifecycleObserver` only. Fragment should not know anything about performance monitoring.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailsTransactionLauncher.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                    performanceTransactionRepository.startTransaction(TRANSACTION_NAME, TransactionOperation.UI_LOAD)
+            }
+            Lifecycle.Event.ON_DESTROY -> {
+                performanceTransactionId?.let {
+                    performanceTransactionRepository.finishTransaction(it, TransactionStatus.ABORTED)
+                }
+            }
+        }
+    }
+}
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailsTransactionLauncher.kt]: This is something I like about this approach. We subscribe to Fragment lifecycle, and if there's a running transaction when `onStop` happens - we abort it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/SpecifyPerformanceMonitoringConfig.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.runBlocking
+import javax.inject.Inject
+
+class SpecifyPerformanceMonitoringConfig @Inject constructor(
+    private val remoteConfigRepository: RemoteConfigRepository,
+    private val analyticsWrapper: AnalyticsTrackerWrapper
+) {
+    operator fun invoke(): PerformanceMonitoringConfig {
+        return runBlocking {
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/util/crashlogging/SpecifyPerformanceMonitoringConfig.kt]: We have to make it blocking, as Sentry initialization happens at the application start. Thankfully, it's just for taking a value from DataStore.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailsTransactionLauncher.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import com.automattic.android.tracks.crashlogging.performance.TransactionStatus
+import dagger.hilt.android.scopes.ViewModelScoped
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@ViewModelScoped
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailsTransactionLauncher.kt:19]: We create this class for every ViewModel. So if Fragment recreates (orientation change), we do not measure the transaction again.

**Comment by wzieba** [comment]
> PR has more than 300 lines of code changing. Consider splitting into smaller PRs if possible.

- we integrate Sentry Performance Integration _and_ add a transaction for Order Details. We could separate it in 2 PRs but it would make the first feature not testable (we wouldn't know if integration works). Other transactions are added in separate PRs
- there's a significant number of unit tests, which I believe shouldn't count in the PR size to review.

So with those 2 arguments, I suggest leaving this PR to review as it is. But if you have other preferences, feel free to ping me!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/config/FirebaseRemoteConfigRepository.kt [hunk 6]
```diff
@@ -87,16 +87,9 @@ class FirebaseRemoteConfigRepository @Inject constructor(
...
             }
 
-    override fun observePerformanceMonitoringSampleRate(): Flow<Double> =
-        observeDoubleRemoteValue(PERFORMANCE_MONITORING_SAMPLE_RATE_KEY)
-            .catch {
-                crashLogging.get().recordException(it)
-                emit(0.0)
-            }
+    override fun getPerformanceMonitoringSampleRate(): Double =
+        remoteConfig.getDouble(PERFORMANCE_MONITORING_SAMPLE_RATE_KEY)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/config/FirebaseRemoteConfigRepository.kt:91]: Explanation of this change: in #7101 we moved to observe Remote Config values in a way, that requires waiting for Remote Config to be initialized (at least execute `setDefaultsAsync`).

As in `SpecifyPerformanceMonitoringConfig` we've been taking the value of sampling in a synchronous way (runBlocking), we had lock:
- `SpecifyPerformanceMonitoringConfig` was blocking thread waiting for value from `FirebaseRemoteConfigRepository`
- `FirebaseRemoteConfigRepository` was waiting for main thread to unblock as it wanted to continue initialization.

We don't want to take `PerformanceMonitoringSampleRate` asynchronously as we need it during app initialization. So I've decided to use a synchronous method, a simple `getDouble` that fixes the problem.

**Review by ThomazFB — APPROVED** [review]
- [X] Check if OrderDetails transaction has been uploaded (up to 1 minute) (**)
- [X] Check if OrderDetails transaction has **not** been uploaded (up to 1 minute) (**)
- [X] Check if OrderDetails transaction has been uploaded but there are less than 10-20 new transactions (up to 1 minute) (**)

Absolutely an impressive work introduced here. I love how Sentry give us so much data to evaluate what is going on. Using the wasabi variant to set a 1.0 sample rate was an amazing idea too. 

I focused on reviewing the output on sentry following your test cases and less on the code itself, reviewing only the main work in the `WooRequestFormatter` and how the `OrderDetailsTransactionLauncher` is configured. I get now your proposed solution of using a ViewModel scoped solution attached to the lifecycle 👍, and nice implementation of the `toSatisfy` rule!

LGTM!

**Comment by wzieba** [comment]
>  I love how Sentry give us so much data to evaluate what is going on. 

I love it too, can't wait to see the data after release 🤞 .

Just for visibility: in the meantime I've added an improvement (https://github.com/woocommerce/woocommerce-android/pull/7069#discussion_r937858237) but it shouldn't have an influence on what you've tested. In fact - it simplified config. I also tested it so I believe we're safe to merge.

**Pull request merged**